pkg/network/ebpf/c/protocols/tls/tls.h

```diff
@@ -31,30 +32,43 @@ typedef struct {
 #define TLS_HANDSHAKE_CLIENT_HELLO 0x01
 #define TLS_HANDSHAKE_SERVER_HELLO 0x02
 
-static __always_inline bool is_tls(const char* buf, __u32 buf_size) {
-    if (buf_size < (sizeof(tls_record_header_t)+sizeof(tls_hello_message_t))) {
-        return false;
+static __always_inline bool is_valid_tls_version(__u16 version) {
+    switch (version) {
+    case SSL_VERSION20:
+    case SSL_VERSION30:
+    case TLS_VERSION10:
+    case TLS_VERSION11:
+    case TLS_VERSION12:
+    case TLS_VERSION13:
+        return true;
     }
 
-    tls_record_header_t *tls_record_header = (tls_record_header_t *)buf;
-    if (tls_record_header->content_type != TLS_HANDSHAKE) {
-        return false;
+    return false;
+}
+
+static __always_inline bool is_tls_handshake(tls_hello_message_t *msg) {
+    switch (msg->handshake_type) {
+    case TLS_HANDSHAKE_CLIENT_HELLO:
+    case TLS_HANDSHAKE_SERVER_HELLO:
+        return is_valid_tls_version(msg->version);
     }
 
-    tls_hello_message_t *tls_hello_message = (tls_hello_message_t *)(buf + sizeof(tls_record_header_t));
-    if (tls_hello_message->handshake_type != TLS_HANDSHAKE_CLIENT_HELLO && tls_hello_message->handshake_type != TLS_HANDSHAKE_SERVER_HELLO) {
+    return false;
+}
+
+static __always_inline bool is_tls(const char *buf, __u32 buf_size) {
+    if (buf_size < (sizeof(tls_record_header_t) + sizeof(tls_hello_message_t))) {
         return false;
     }
 
-    switch (tls_hello_message->version) {
-        case SSL_VERSION20:
-        case SSL_VERSION30:
-        case TLS_VERSION10:
-        case TLS_VERSION11:
-        case TLS_VERSION12:
-        case TLS_VERSION13:
-            return true;
+    tls_record_header_t *tls_record_header = (tls_record_header_t *)buf;
+    switch (tls_record_header->content_type) {
+    case TLS_HANDSHAKE:
+        return is_tls_handshake((tls_hello_message_t *)(buf + sizeof(tls_record_header_t)));
+    case TLS_APPLICATION_DATA:
+        return is_valid_tls_version(tls_record_header->version);
```

Comment:
It would be a good idea to check the chunk length of the payload as well to avoid false positive
something like `tls_record_header->length == buf_size`

Reply:
Also, check the length is not longer than the limit of 2^14

Reply:
> Also, check the length is not longer than the limit of 2^14

Added.

> It would be a good idea to check the chunk length of the payload as well to avoid false positive
something like tls_record_header->length == buf_size

Currently, `is_tls` is only given the size of the classification fragment (which is 24 bytes). We could modify it to take the skb size, but do we have the guarantee that a skb can only contain a single TLS record?